### ISSUE TITLE
Variable hyperparameter results!

### ISSUE THREAD

**Issue by ghaffari903** [AI4Finance-Foundation/FinRL#440]
I do not know why the results of optimizing the hyperparameters are different each time (30 times each time)!
How can we get fixed  optimized hyperparameters values?
I dont find deterministic(False or True)

**Comment by XiaoYangLiu-FinRL** [comment]
Please give clear reference to code file, and code lines. This would help a lot.

**Comment by skhonsha** [comment]
thank you,
https://github.com/AI4Finance-Foundation/FinRL-Meta/blob/master/Demo_FinRL_Meta_Integrate_Trends_data_to_DOW_Jones.ipynb

some results:

{'batch_size': 64, 'n_steps': 512, 'learning_rate': 0.005113164177180871, 'ent_coef': 4.16124607533727e-07}

{'batch_size': 256, 'n_steps': 128, 'learning_rate': 0.23839412890469167, 'ent_coef': 1.749555970832671e-07}

{'batch_size': 64, 'n_steps': 256, 'learning_rate': 0.004226282022135361, 'ent_coef': 2.140112929396758e-05}

**Comment by ghaffari903** [comment]
thank you,
https://github.com/AI4Finance-Foundation/FinRL-Meta/blob/master/Demo_FinRL_Meta_Integrate_Trends_data_to_DOW_Jones.ipynb

some results:

{'batch_size': 64, 'n_steps': 512, 'learning_rate': 0.005113164177180871, 'ent_coef': 4.16124607533727e-07}

{'batch_size': 256, 'n_steps': 128, 'learning_rate': 0.23839412890469167, 'ent_coef': 1.749555970832671e-07}

{'batch_size': 64, 'n_steps': 256, 'learning_rate': 0.004226282022135361, 'ent_coef': 2.140112929396758e-05}

**Comment by Athe-kunal** [comment]
What agent are you using?
If you are using Stable Baselines3, then I just committed deterministic parameter to DRL_prediction [here](https://github.com/AI4Finance-Foundation/FinRL/blob/master/finrl/drl_agents/stablebaselines3/models.py#L110). So you can have deterministic results in the testing or validation phase. Please try again, let me know if you are able to get the same results

**Comment by ghaffari903** [comment]
THANK YOU, yes Stable Baselines3, HOW ABOUT elegantrl and rllib?

**Comment by ghaffari903** [comment]
and is the same for  DRL_prediction_load_from_file?

**Comment by Athe-kunal** [comment]
Yes, you can do a similar thing for DRL_prediction_load_from_file

**Comment by ghaffari903** [comment]
![vv](https://user-images.githubusercontent.com/97940416/150651824-e0baa324-7b90-46cb-8282-ad963268a89e.png)

**Comment by Athe-kunal** [comment]
I have made the changes in the models.py file of SB3, please refer it

**Comment by ghaffari903** [comment]
Thank you very much, do you have any optimizaion on drlelegant and rllib to compare with Stable Baselines3?

**Comment by Athe-kunal** [comment]
Yes, I have Ray tune for RLlib models, you can refer [here](https://github.com/AI4Finance-Foundation/FinRL/blob/master/FinRL_Raytune_for_Hyperparameter_Optimization_RLlib%20Models.ipynb)

**Comment by ghaffari903** [comment]
Hi Athe,
result after use deterministic option (It is still variable):

{'batch_size': 512, 'n_steps': 128, 'learning_rate': 0.044466289554754464, 'ent_coef': 0.00029033694281285587}

6.663615733011714

 

Successfully load model tuned_models/ppo_dict_values([512, 128, 0.044466289554754464, 0.00029033694281285587])

episode_return 0.7025797771208984

 
--------------------------------------------------------------------------------------------------
{'batch_size': 128, 'n_steps': 1024, 'learning_rate': 0.034702669886504146, 'ent_coef': 1.3934502251337566e-08}

7.085039707694913

 

Successfully load model tuned_models/ppo_dict_values([128, 1024, 0.034702669886504146, 1.3934502251337566e-08])

episode_return 0.5851003146268554

**Comment by Athe-kunal** [comment]
Please try passing a seed value to the model while training for reproducible results. You can pass the seed value in the get_model attribute. Refer [here](https://github.com/AI4Finance-Foundation/FinRL/blob/master/finrl/drl_agents/stablebaselines3/models.py#L75)

**Comment by ghaffari903** [comment]
> 

deterministic must be true?

**Comment by Athe-kunal** [comment]
Yes for reproducible results you need to set deterministic=True.

**Comment by ghaffari903** [comment]
both train and test?

**Comment by Athe-kunal** [comment]
For test and validation set

**Comment by ghaffari903** [comment]
Dear @Athe-kunal The answers are always variable! I have another question, can one of the models that worked very well be improved in the another training notebook? That is, retraining the model?

**Comment by Athe-kunal** [comment]
Oh, they are still variable. Okay here I am sharing a link which also tries to reproduce results at CuDNN level. [tutorial DQN](https://github.com/Curt-Park/rainbow-is-all-you-need/blob/master/01.dqn.ipynb)

Here you can scroll down to see that he used torch.cudnn =determinstic. I will see later how to integrate to FinRL.

**Comment by ghaffari903** [comment]
thank you so much, I wish it was an communicating group that users could tap and exchange the results of the models on different data with different configurations.

**Comment by Athe-kunal** [comment]
Yeah, I agree but the the hyperparameters tuning part is very sensitive and varies on lot of configuration. So it is hard to reproduce results. Do stress test the system to find a good set of hyperparameters and test them further. Empirical evidences are helpful in validating these ideas
Anyway, if you don't have any further queries, I will close this issue and please feel free to reopen it.
